### PR TITLE
Introduce a gate/check GHA job

### PR DESCRIPTION
**Pull request by webknjaz**
This adds a GHA job that reliably determines if all the required dependencies have succeeded or not.

It also allows to reduce the list of required branch protection CI statuses to just one — `check`. This reduces the maintenance burden by a lot and have been battle-tested across a small bunch of projects in its action form and in-house implementations of other people.

This action is now in use in aiohttp (and other aio-libs projects), CherryPy, some of the Ansible repositories, pip-tools, all of the jaraco's projects (like `setuptools`, `importlib_metadata`), some of the hynek's projects (like `attrs`, `structlog`), some PyCQA, PyCA, PyPA and pytest projects, a few AWS Labs projects. Admittedly, I maintain a few of these but it seems to address some of the pain folks have: https://github.com/jaraco/skeleton/pull/55#issuecomment-1106638475.

I saw a few attempts to reduce the maintenance burden for GHA and figured
this might be useful for CPython too, which is why I'm submitting this patch.
Looking forward to hearing what you think!

The story behind this is explained in more detail at https://github.com/marketplace/actions/alls-green#why.

### CODE REVIEW

File: .github/workflows/build.yml
Comment (webknjaz) [inline .github/workflows/build.yml]: By default, all the jobs listed in the `check` job's `needs` section are required to succeed in terms of branch protection. This setting emulates having the job non-required (in combination with the check job below being the only thing required in branch protection).

File: .github/workflows/build.yml
Comment (webknjaz) [inline .github/workflows/build.yml:525]: This is mandatory because otherwise it'll get a `skipped` status in some cases and the branch protection sees that as `skipped == success` which is undesired.

File: .github/workflows/build.yml
Comment (webknjaz) [inline .github/workflows/build.yml:550]: This setting tells the action that if the listed jobs (comma-separated) got skipped, that's okay. But the jobs are allowed to be skipped based on the same condition they've got set. If the first job requests test runs, this list will compute as empty and none of the jobs will be allowed to be skipped (meaning that skips would turn into failures).

**Comment by hugovk** [comment]
Thanks for the PR!

Some points to consider: 

1. This changes the security levels: a small group of [6 org owners + 8 repo admins](https://devguide.python.org/developer-workflow/development-cycle/#repository-administration) can change the required checks via repo settings. With this PR, any of the ~90 active core devs can merge changes to this file. Is that okay? I think it's probably fine, we're not deploying or anything from here, but let's make sure.

2. Whilst this does reduce maintenance burden in administering which checks are required or not, is that something that needs changing often?
   For projects that test a big matrix of Python versions, it's definitely a plus: no need to add/remove version-specific checks via the interface once or twice a year. Maybe less so here?

3. Has this ~14 vs. ~90 been a bottleneck in the past?

**Comment by webknjaz** [comment]
@hugovk these are good points and I haven't considered some of them.

1. I should probably document the security considerations in README. It seems like this might be solvable by requiring reviews from code owners, for example. This should limit the scope of the CI updates made flying under the radar. It also occurred to me that the repository settings updates happen unnoticed, with no traceability for the larger public — only the admins would be able to see this in the security audit log and only if they specifically search for such changes. Whereas with the branch-protection-as-a-code, it's all there, verifiable via https://github.com/python/cpython/commits/main/.github/workflows/build.yml. 
2. I've been offering similar PRs to many projects with matrixes of various sizes, and I've seen the maintainers forgetting to add new matrix factors. After all, it's a mental overhead having to memorize to perform some extra action items far from the place where you're editing the source code, especially since they aren't even in the code.
3. That's not something I can observe/verify :man_shrugging: — maybe @ezio-melotti knows?

File: .github/workflows/build.yml
Comment (ambv) [inline .github/workflows/build.yml]: What's the reason this is pinned to v1.2.1 instead of using the release name? Also, I looked at the repo of the action and the checks are red there, is that expected?

Comment (webknjaz) [reply .github/workflows/build.yml]: > What's the reason this is pinned to v1.2.1 instead of using the release name?

No particular reason from my side but when I contributed this to other projects, some wanted to have commit sha pinning. Dependabot is able to bump such commit hash based pins so it's usually considered the best of both worlds.

For myself, I usually do this:
```suggestion
      uses: re-actors/alls-green@release/v1
```
But I appreciate that some people want better traceability for non-GH-official actions. It's up to you really.

If you choose to go for the `release/v1` branch, it represents stable API. Note that Dependabot will not be updating the reference in this case.

> Also, I looked at the repo of the action and the checks are red there, is that expected?

Yes, kinda. It is necessary for internal smoke-testing. Some of the jobs are red but the check job isn't. The overall workflow status may be red but it's not necessary a reflection of the gate check, it's just a quirk of how GitHub represents statuses — it doesn't know that we actually care about the outcome of one single job which decides internally whether the dependencies are allowed to fail or be skipped, and if any of the workflow jobs is red, it marks the whole thing as red. The branch protection is a disconnected concept and is not applied in all possible contexts, which is why GH can't make assumptions automatically, and it doesn't provide a way to declare what's important on the GHA declaration level.
It's weird but we can't do anything about it, unless GitHub decides to make updates...

Comment (ambv) [reply .github/workflows/build.yml]: Alright, all this makes sense to me.

**Comment by ambv** [comment]
I don't know if this is worth it for CPython as a number of jobs we require are outside of GitHub Actions (CLA, Issue Number, and News). So this check will increase complexity but won't actually solve the problem of bringing all required checks in one place.

**Comment by zware** [comment]
Agreed with @ambv: this looks like significant added complexity for not much actual gain in this project.  It looks like a nice idea for some other setups, but doesn't feel like a great fit for us here.  It's also possible I'm just missing the point, though :)

**Comment by brettcannon** [comment]
> > 1. I should probably document the security considerations in README. It seems like this might be solvable by requiring reviews from code owners, for example.
> 
> Just to note, that would be a pretty _huge_ change in terms of workflow for this repo, since `CODEOWNERS` is used here for much more than just who is _required_ to review certain changes.

FYI the SC has explicitly stated individuals cannot act as owners on any one piece of code (i.e. we **all** own the code), so requiring `CODEOWNERS` approval is a non-starter IMO.

**Comment by ambv** [comment]
Alright, so I'm closing this based on the latest feedback. Thanks for putting this together, @webknjaz, and sorry it wasn't a good match for us at this time!

**Comment by ambv** [comment]
Given that our requirements changed, with Auto-merge enabled, and more checks existing now, I would like to reconsider this PR for inclusion.

@webknjaz, would you be so kind as to adapt this to the current state of `main`?

**Comment by webknjaz** [comment]
@ambv so I just realized that the docs are its own workflow. The only way this is gonna work is wiring that in as a reusable workflow.

File: .github/workflows/build.yml
Comment (ambv) [inline .github/workflows/build.yml]: "All required checks pass"

Comment (ambv) [reply .github/workflows/build.yml]: Job ID: `s/check/all-required-green/`

**Comment by webknjaz** [comment]
A quick summary of our in-person interaction. Łukasz and I agreed to implement the following:
- rename the `check` job ID and name
- convert the docs workflow into a reusable one
- integrate that into the main `build` workflow, making use of the dynamic change detection that's already present there
- move the "allowed failures" declarations to the check

@ambv do we need to do the same to the `build_msi` workflow?

File: .github/workflows/build.yml
Comment (webknjaz) [reply .github/workflows/build.yml]: Patched. Marking the conversation as resolved.

File: .github/workflows/build.yml
Comment (webknjaz) [reply .github/workflows/build.yml]: This has been moved to the action call. Resolving.

**Comment by AA-Turner** [comment]
@webknjaz would you be able to explain a little about 'reusable workflow' for the documentation workflow? I don't know what makes it a special case, or what a reusable workflow is!

A

File: .github/workflows/reusable-docs.yml
Comment (AA-Turner) [inline .github/workflows/reusable-docs.yml]:
```suggestion
name: Docs
```

Comment (AA-Turner) [reply .github/workflows/reusable-docs.yml]: (I find the intented meaning of emoji harder to recognise at a glance, especially when I'm switching contexts -- words are much more explicit!)

Comment (webknjaz) [reply .github/workflows/reusable-docs.yml]: This is specifically done to make the UI displayed things shorter. So I'll wait for other people's feedbacks to see if it makes sense to revert to the previous name.
The thing is that with reusable workflows, the strings will be longer. So I kept the internal job name, but not the outer layer things.

**Comment by webknjaz** [comment]
> explain a little about 'reusable workflow' for the documentation workflow? I don't know what makes it a special case, or what a reusable workflow is!

TL;DR it's a way to include one workflow into another: https://docs.github.com/en/actions/using-workflows/reusing-workflows.

Łukasz and I talked at the PyCon Sprints and I explained in person how it's useful for this case. Basically, you cannot depend on jobs running in separate workflows, so this patch makes it so the docs workflow is included inside the main one and doesn't get executed as a standalone thing. It is a nice way of organizing bits of automation without having to shove everything in the same file. Reusable workflows can also exist in external repositories, but that's not our case. I did show other workarounds for making auto-merge work, but this one is the most elegant solution.

Please, check out this graph to see what this looks like: https://github.com/python/cpython/actions/runs/4814439259.

**Comment by webknjaz** [comment]
> @ambv do we need to do the same to the `build_msi` workflow?

And a follow-up question: should I submit the first commit (the reusable workflow intro) as a separate PR so that it's more or less atomic?

**Comment by webknjaz** [comment]
@AA-Turner oh, and in addition to the above I suggested Łukasz **_not to add_** the new job to the branch protection right away but rather observe and allow for some grace period before going full-in. So this way it won't introduce any friction or side effects to people's experiences.

**Comment by AA-Turner** [comment]
To confirm my understanding: the 'all-required-green' job works iff all required jobs are within a single workflow.

I suppose my challenge would be extensibility for the future, as this might make it awkward to add more jobs. Say in a hypothetical world we want to require that the [verify wheels](https://github.com/python/cpython/blob/main/.github/workflows/verify-ensurepip-wheels.yml) action passes on every PR -- if I'm correct this would require re-writing the action and including it within the ``build.yml`` file. Could you confirm my understanding?

Is there a way to keep the benefits of the 'configuration as code' approach that this takes whilst still maintaining distinct workflows (for extensibility purposes), perhaps via uploading a dummy artefact or etc to a well-known location on GHA or some other way of sharing state?

A

**Comment by ambv** [comment]
Not sure about `build_msi`; I would leave it be for now.

You could make the reusable intro a separate PR to make reviewing this one easier.

**Comment by webknjaz** [comment]
Hey @AA-Turner, thanks for the questions! I hope my answers below will be able to shed some light:

> To confirm my understanding: the 'all-required-green' job works iff all required jobs are within a single workflow.

Yes. Alternatively, there could be many workflows with disconnected matrixes and each would need such a job (though, with different names!) — every each of them should need to be added to the branch protection.

This works by adding jobs as dependencies via `needs:` which is not possible across workflows.

> I suppose my challenge would be extensibility for the future, as this might make it awkward to add more jobs. Say in a hypothetical world we want to require that the [verify wheels](https://github.com/python/cpython/blob/main/.github/workflows/verify-ensurepip-wheels.yml?rgh-link-date=2023-04-27T00%3A40%3A04Z) action passes on every PR -- if I'm correct this would require re-writing the action and including it within the `build.yml` file. Could you confirm my understanding?

Kinda. I'd like to clarify that it's nicer to keep things in different YAML files. And the inclusion would just be a one-liner with a `uses:`, not physically copying the contents over. The result would be that the definition lives in its own dedicated YAML file, while being a part of the common workflow in the runtime (test time?).
A unified workflow also allows making use of various change detection techniques, ones and calling some workflows conditionally.
I would even argue that most of the jobs that are currently declared in the `build.yml` workflow file should move out at some point, for better separation of different workflow layers. Though, this is definitely a separate discussion as I want to keep the PR atomic, focused on a single thing, without causing a scope creep.

> Is there a way to keep the benefits of the 'configuration as code' approach that this takes whilst still maintaining distinct workflows (for extensibility purposes), perhaps via uploading a dummy artefact or etc to a well-known location on GHA or some other way of sharing state?

I'm not sure what you're asking here. But as mentioned above, I'd very much recommend splitting out things and keeping on the high-level inclusion and conditionals in the main one. If you're asking about sharing artifacts across jobs in general, that is certainly possible with the “stdlib” actions/upload-artifact and actions/download-artifacts. The workflows could refer to the same artifact names. Also, the reusable workflows can accept inputs, just like actions, so the caller could pass some “arguments” into the called workflows (think of in like calling functions, in a way).

**Comment by webknjaz** [comment]
@ambv I'll submit the PR shortly. In the meanwhile, I noticed a bug in the docs workflow that has existed there for a while and is being ignored almost silently — it's here https://github.com/python/cpython/actions/runs/4814439259/jobs/8572091258#step:9:3. There are two problems with that — `touch` does not get any args so it produces an error output. And a multiline shell script executed without `set -x` that keeps running after one of the commands in the middle failed, and the final outcome is the result of the very last instruction in that script.

**Comment by webknjaz** [comment]
The separate PR is here: https://github.com/python/cpython/pull/103914.

**Comment by webknjaz** [comment]
> Say in a hypothetical world we want to require that the [verify wheels](https://github.com/python/cpython/blob/main/.github/workflows/verify-ensurepip-wheels.yml?rgh-link-date=2023-04-27T00%3A40%3A04Z) action passes on every PR -- if I'm correct this would require re-writing the action and including it within the `build.yml` file.

I realized that I didn't fully respond to this since I didn't understand what you imply. I suppose you were also curious about conditional runs/change detection. That would indeed move to the main workflow. Though, the change detection itself can be separated out too, if needed. Either as a reusable workflow or as a composite action. We can solve this elegantly when the time comes.

File: .github/workflows/reusable-docs.yml
Comment (hugovk) [reply .github/workflows/reusable-docs.yml]: Let's keep "Docs", it's short enough already.

By the way, we usually don't force push to https://github.com/python/ repos, it makes it a bit harder to review, and everything gets squash merged anyway. Thanks!

PS it was great to meet you at PyCon!

Comment (AlexWaygood) [reply .github/workflows/reusable-docs.yml]: I agree with Adam and Hugo, I strongly prefer words over emojis, especially when they're short! :)

Comment (webknjaz) [reply .github/workflows/reusable-docs.yml]: Please confirm that you are fine seeing a multi-segment thing like `Tests / Docs / Docs` in the UI with duplicate parts. I'm not sure you fully understand the implication of how this influences the check names.

Comment (webknjaz) [reply .github/workflows/reusable-docs.yml]: > By the way, we usually don't force push to https://github.com/python/ repos, it makes it a bit harder to review, and everything gets squash merged anyway.

Yes, sometimes it makes it easier but I'll need to rebase after the separate PR is merged anyway. Plus it makes me easier to smuggle a nice commit message into the resulting thing.

Comment (AlexWaygood) [reply .github/workflows/reusable-docs.yml]: > Please confirm that you are fine seeing a multi-segment thing like `Tests / Docs / Docs` in the UI. I'm not sure you fully understand the implication of how this influences the check names.

I think I'd be fine with having `Tests / Docs / Docs` in the list of checks at the bottom of the PR if it means we can avoid this, when you click through to view the results of the checks:

6E362A53-9461-4E8B-89B8-8037B1770EE9

Comment (webknjaz) [reply .github/workflows/reusable-docs.yml]: Oh, so you're talking about a different view. That's fair. I'd like to hear from others and @llanga but now I at least understand where this is coming from...

**Comment by hugovk** [comment]
> @ambv I'll submit the PR shortly. In the meanwhile, I noticed a bug in the docs workflow that has existed there for a while and is being ignored almost silently — it's here [python/cpython/actions/runs/4814439259/jobs/8572091258#step:9:3](https://github.com/python/cpython/actions/runs/4814439259/jobs/8572091258#step:9:3). There are two problems with that — `touch` does not get any args so it produces an error output. And a multiline shell script executed without `set -eEuo pipefail` that keeps running after one of the commands in the middle failed, and the final outcome is the result of the very last instruction in that script.

Hmm, in that case there's no modified docs in the PR, so nothing to `touch` and rebuild. It could be nicer and skip when there are no files.

But more importantly, since this PR was opened we've refactored it to touch via a Python script instead of the `touch` command, so the behaviour may have changed too. Something to check in the other PR.

File: .github/workflows/reusable-docs.yml
Comment (hugovk) [reply .github/workflows/reusable-docs.yml]: I'm fine with:

* Tests / Docs / Docs
* Tests / Docs / Docs (Oldest Sphinx)
* Tests / Docs / Doctest

We can maybe rename that final part later, but for now I prefer to keep words.

(And it's @ambv not `@llanga` on GitHub :)

Comment (webknjaz) [reply .github/workflows/reusable-docs.yml]: Oh, thanks for correcting me 🤦‍♂️

Comment (hugovk) [reply .github/workflows/reusable-docs.yml]: > Yes, sometimes it makes it easier but I'll need to rebase after the separate PR is merged anyway. Plus it makes me easier to smuggle a nice commit message into the resulting thing.

The nice commit message is lost when we squash merge, so don't worry about it. And we can merge `main` into here when the other PR is merged, no need to rebase.

Shall we mark this PR as draft and continue on the other one?

Comment (webknjaz) [reply .github/workflows/reusable-docs.yml]: > We can maybe rename that final part later, but for now I prefer to keep words.

Note that this specific line refers to the middle part, not the final one.

Comment (webknjaz) [reply .github/workflows/reusable-docs.yml]: > The nice commit message is lost when we squash merge, so don't worry about it. And we can merge main into here when the other PR is merged, no need to rebase.

Well, it's a habit I cultivate and have to keep doing because habits don't work if you apply them selectively. This is also the reason I don't like squashes since they ruin history in certain ways.

> Shall we mark this PR as draft and continue on the other one?

Yeah, I was considering this. I've been basically waiting for Łukasz to finish his journey back home to resume the work on this. I'll be leaving SLC in about 12 hours myself and probably won't get to the laptop until the Saturday evening (Prague time).

File: .github/workflows/build.yml
Comment (webknjaz) [inline .github/workflows/build.yml:527]: TODO: add hypothesis here too

**Comment by webknjaz** [comment]
@ambv @hugovk I think this is ready.

**Comment by webknjaz** [comment]
@ambv @hugovk let's see if this is ready now...

File: .github/workflows/build.yml
Comment (hugovk) [inline .github/workflows/build.yml]: Let's remove the emoji, it's a little odd to see a double green (and we could sometimes see a "❌✅ All required checks pass"):

image


```suggestion
    name: All required checks pass
```

File: .github/workflows/build.yml
Comment (hugovk) [inline .github/workflows/build.yml]:
```suggestion
    - name: Check whether the needed jobs succeeded or failed
```

File: .github/workflows/build.yml
Comment (hugovk) [reply .github/workflows/build.yml:550]: Do I understand this right?

- If `check-docs` starts, it must pass; if they skip, that's fine.

- If `check_generated_files` or any of the other jobs in its group start, they must all pass; if they all skip, that's fine.

- If `test_hypothesis` starts, it must pass; if they skip, that's fine.

File: .github/workflows/build.yml
Comment (webknjaz) [reply .github/workflows/build.yml]: Done!

File: .github/workflows/build.yml
Comment (webknjaz) [reply .github/workflows/build.yml]: Done!

File: .github/workflows/build.yml
Comment (webknjaz) [reply .github/workflows/build.yml:550]: @hugovk almost: this is paired with the setting above :point_up: — essentially, if something is allowed to be skipped, and it is skipped, the action will treat it as non-failure.
But if something is allowed to be skipped and is not skipped, whether its failure is to be treated as such is controlled by the `allowed-failures` input.

Also, not they are not grouped inside the action, the templating here just produces a comma-separated list of things and the action doesn't know it was in some group externally. So yes, such a relation kinda exists, but I would say that it's indirect.

And since there's no “groups”, if any of the jobs that are allowed to be skipped run, each individual job is evaluated separately from the “group”, based on the fact of failures being allowed or not.

N.B. Whether a job is allowed to fail is also contributed by the `continue-on-error` setting, which is mostly useful for controlling individual matrix jobs. If a job was allowed to fail prior to this one, the action will not “see” if it was red and treat it as green.

Comment (webknjaz) [reply .github/workflows/build.yml:550]: > * If `test_hypothesis` starts, it must pass; if they skip, that's fine.

With the context I added earlier, seeing that `test_hypothesis` is listed in `allowed-failures` unconditionally, it will never cause failures. This reflects the current branch protection configuration as seen in the Checks widget in PRs.

**Review by hugovk — APPROVED** [review]
Thanks! Let's give @ambv some time to check before merging.

**Comment by webknjaz** [comment]
@hugovk IIRC Łukasz and I talked about merging this PR but not changing the branch protection settings for a while so the check name would start being reported in the Checks widget on old PR updates, otherwise they'll get blocked until updated.

**Comment by hugovk** [comment]
Thanks again!

**Comment by webknjaz** [comment]
UPD: this has been backported to 3.12 and 3.11 and the branch protection changed accordingly.

**Pull request merged**
